sbi/inference/snpe/snpe_a.py

```diff
@@ -351,3 +365,558 @@ def _expand_mog(self, eps: float = 1e-5):
+        with
+        $c_k = logdet(S_k) - logdet(S_0) - logdet(S_k^\prime) +
+             + m_k^T P_k m_k - m_0^T P_0 m_0 - m_k^\prime^T P_k^\prime m_k^\prime$
+        (see eqs. (25, 26) in Appendix C of [1])
+
+        Args:
+            means_post: Means of the posterior.
+            precisions_post: Precision matrices of the posterior.
+            covariances_post: Covariance matrices of the posterior.
+            logits_pprior: Component weights (i.e. logits) of the proposal prior.
+            means_pprior: Means of the proposal prior.
+            precisions_pprior: Precision matrices of the proposal prior.
+            logits_d: Component weights (i.e. logits) of the density estimator.
+            means_d: Means of the density estimator.
+            precisions_d: Precision matrices of the density estimator.
+
+        Returns: Component weights of the proposal posterior.
+        """
+
+        num_comps_pprior = precisions_pprior.shape[1]
+        num_comps_d = precisions_d.shape[1]
+
+        # Compute the ratio of the logits similar to eq (10) in Appendix A.1 of [2]
+        logits_pprior_rep = logits_pprior.repeat_interleave(num_comps_d, dim=1)
+        logits_d_rep = logits_d.repeat(1, num_comps_pprior)
+        logit_factors = logits_d_rep - logits_pprior_rep
+
+        # Compute the log-determinants
+        logdet_covariances_post = torch.logdet(covariances_post)
+        logdet_covariances_pprior = -torch.logdet(precisions_pprior)
+        logdet_covariances_d = -torch.logdet(precisions_d)
+
+        # Repeat the proposal and density estimator terms such that there are LK terms.
+        # Same trick as has been used above.
+        logdet_covariances_pprior_rep = logdet_covariances_pprior.repeat_interleave(
+            num_comps_d, dim=1
+        )
+        logdet_covariances_d_rep = logdet_covariances_d.repeat(1, num_comps_pprior)
+
+        log_sqrt_det_ratio = 0.5 * (  # similar to eq (14) in Appendix A.1 of [2]
+            logdet_covariances_post
+            + logdet_covariances_pprior_rep
+            - logdet_covariances_d_rep
+        )
+
+        # Compute for proposal, density estimator, and proposal posterior:
+        exponent_pprior = utils.batched_mixture_vmv(
+            precisions_pprior, means_pprior  # m_0 in eq (26) in Appendix C of [1]
+        )
+        exponent_d = utils.batched_mixture_vmv(
+            precisions_d, means_d  # m_k in eq (26) in Appendix C of [1]
+        )
+        exponent_post = utils.batched_mixture_vmv(
+            precisions_post, means_post  # m_k^\prime in eq (26) in Appendix C of [1]
+        )
+
+        # Extend proposal and density estimator exponents to get LK terms.
+        exponent_pprior_rep = exponent_pprior.repeat_interleave(num_comps_d, dim=1)
+        exponent_d_rep = exponent_d.repeat(1, num_comps_pprior)
+        exponent = -0.5 * (
+            exponent_d_rep - exponent_pprior_rep - exponent_post  # eq (26) in [1]
```

Comment:
I agree that this is now exactly as in eq (26) in [1]. Now I have a question. We compared this to eq (14) and (10) in the SNPE-C paper [2], and thought that basically the normals of the proposal and the prior change place  in (10). However, we forgot that SNPE-A is estimating a different quantity. I am surprised that the test examples worked so well with the wrong `exponent` :smile:

Reply:
All tests in `sbi` are unimodal. Since this bug only affects the logits, it will not show in these cases. I found the bug by explicitly writing a test to check the division performed in SNPE-A. The test is also in this PR (see the commit "added unit test for snpe-a").

Reply:
Ah okay, this now makes sense to me. Thanks for clarifying. I hope this was the only bug.